platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/GraphVariantSelector.java

```diff
@@ -276,21 +280,41 @@ private ImmutableList<VariantGraphResolveState> filterVariantsByRequestedCapabil
      */
-    private MatchResult providesAllCapabilities(ComponentGraphResolveMetadata targetComponent, Collection<? extends Capability> explicitRequestedCapabilities, ImmutableCapabilities providerCapabilities) {
+    private static MatchResult providesAllCapabilities(
+        ComponentGraphResolveMetadata targetComponent,
+        Set<CapabilitySelector> explicitRequestedCapabilities,
+        ImmutableCapabilities providerCapabilities
+    ) {
         ImmutableSet<ImmutableCapability> providerCapabilitiesSet = providerCapabilities.asSet();
+        ModuleIdentifier targetModuleId = targetComponent.getModuleVersionId().getModule();
         if (providerCapabilitiesSet.isEmpty()) {
             // producer doesn't declare anything, so we assume that it only provides the implicit capability
             if (explicitRequestedCapabilities.size() == 1) {
-                Capability requested = explicitRequestedCapabilities.iterator().next();
-                ModuleVersionIdentifier mvi = targetComponent.getModuleVersionId();
-                if (requested.getGroup().equals(mvi.getGroup()) && requested.getName().equals(mvi.getName())) {
-                    return MatchResult.EXACT_MATCH;
+                CapabilitySelector requested = explicitRequestedCapabilities.iterator().next();
+                if (requested instanceof SpecificCapabilitySelector) {
+                    SpecificCapabilitySelector exactSelector = (SpecificCapabilitySelector) requested;
+                    if (exactSelector.getGroup().equals(targetModuleId.getGroup()) && exactSelector.getName().equals(targetModuleId.getName())) {
+                        return MatchResult.EXACT_MATCH;
+                    }
                 }
+                // A feature capability selector should never match the default capability of a component.
             }
         }
-        for (Capability requested : explicitRequestedCapabilities) {
-            String requestedGroup = requested.getGroup();
-            String requestedName = requested.getName();
+        for (CapabilitySelector requested : explicitRequestedCapabilities) {
+            String requestedGroup;
+            String requestedName;
+            if (requested instanceof SpecificCapabilitySelector) {
+                SpecificCapabilitySelector specificSelector = (SpecificCapabilitySelector) requested;
+                requestedGroup = specificSelector.getGroup();
+                requestedName = specificSelector.getName();
+            } else if (requested instanceof FeatureCapabilitySelector) {
+                FeatureCapabilitySelector featureSelector = (FeatureCapabilitySelector) requested;
+                requestedGroup = targetModuleId.getGroup();
+                requestedName = targetModuleId.getName() + "-" + featureSelector.getFeatureName();
+            } else {
+                throw new IllegalStateException("Unknown capability selector type: " + requested.getClass());
+            }
```

Comment:
🤔 There are several blocks that switch off the specific sub-type of `CapabilitySelector`.  I don't have an idea for restructuring this hierarchy, but this seems like an OOP smell to me.

Reply:
There are a number of places we do this `switch off`:

- ModuleMetadataSpecBuilder - GMM does not support feature selectors, but even if it did we need to know which selector type to write to JSON.
- CapabilitySelectorSerializer - This seems like an appropriate place to `switch`
- AbstractExternalModuleDependency, DefaultProjectDependency, DefaultModuleComponentSelector, DefaultProjectComponentSelector,  - The `switch` is to implement legacy support for `getRequestedCapabilities`
- GraphVariantSelector - This is where we actually need to know which selector is used so we can match against the actual capabilities of the target variant. This seems appropriate to `switch` on
- ResolutionFailureHandler - This is related to another discussion and we can and should get rid of this (perhaps in this PR, perhaps later)

In practice there are a couple use cases for doing this:
- Serializing (to GMM, to dependency cache) These are appropriate places to switch IMO
- Legacy support for `getRequestedCapabilities`. Each of these implementations are slightly different as they have different inputs. I don't see this code as long-lived as `getRequestedCapabilities` will either be removed in 9.0 or 10.0
- Performing capability selection. This seems appropriate to switch IMO
- ResolutionFailureHandler - We can likely avoid the switch here, but TBD based on the type of logic each failure handler needs to implement.

Reply:
Does it make sense to add a `boolean matches(Capability)` here to the public type?

Reply:
The discussion Justin and I had about this was about whether or not the idea of a `CapabilitySelector` should define some behavior related to its purpose and usable to select a capability without knowing about its subtypes.

`ComponentSelector`, which is a similar concept, defines a `boolean matchesStrictly(ComponentIdentifier identifier)` method.  

We wouldn't expect user code to ever have a need to use any behavior we define on `CapabilitySelector`.  Dependency resolution, however, does use some behavior, it use gets to that behavior via switching on `instanceof` to determine what that behavior should be.  We don't expect to need to ban users from calling this behavior, just that it would never be actually useful to them.

We would also expect to be able to convert this selector hierarchy to a sealed type hierarchy in the relatively near future.